File: LeetCode/滑动窗口/209.长度最小的子数组.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int hh = 0, tt =-1, sum = 0, l = nums.size();
        int q[100005];
        for(int i = 0;i <= nums.size();i++) {
            while(hh <= tt && sum >= target) {
                l = min(l, tt-hh + 1);
                sum -= q[hh];
                hh++;
            }
            if(i == nums.size())
                break;
            q[++tt] = nums[i];
            sum += nums[i];
        }
        if (hh == 0)
            return 0;
        return l;
    }
};

File: LeetCode/滑动窗口/209.长度最小的子数组.py
#
# @lc app=leetcode.cn id=209 lang=python3
#
# [209] 长度最小的子数组
#

# @lc code=start
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        ans = n+1
        s = 0
        left = 0
        # 枚举右端点写法
        for right, x in enumerate(nums):
            s += x
            while s - nums[left] >= target:
                s -= nums[left]
                left += 1
            if s >= target:
                ans = min(ans, right-left+1)
        return  ans if ans <= n else 0 
# @lc code=end

#
# @lc app=leetcode.cn id=209 lang=python3
#
# [209] 长度最小的子数组
#

# @lc code=start
# 第二种写法
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        ans = n+1
        s = 0 
        left = 0
        for right, x in enumerate(nums):
            s += x
            # 单调性
            while s >= target:
                ans = min(ans, right-left+1)
                s -= nums[left]
                left += 1 # 至多是n次
        return  ans if ans <= n else 0 
# @lc code=end

